File: Problem11_ContainerWithMostWater.cpp
#include <bits/stdc++.h>
using namespace std;
/*
* Ý tưởng:
* Bắt đầu với cột đầu tiên i=0, và cột cuối cùng j=height.size - 1
* cố định cột có height nhỏ hơn, giả sử là cột j, thể tích = h(j)*(j-i)
* bây giờ nếu ta duyệt i = 1 tới height.size - 2
* thể tích = min(h(i),h(j))*(j-i)
* rõ ràng j-i giảm vì giữ nguyên i và tăng j, còn min(h(i),h(j)) <= h(j)
* => giá trị thể tích luôn nhỏ hơn h(j)*(j-i) -> chỉ cần nhảy j = j - 1
*/
class Solution {
public:
    int maxArea(vector<int>& height) {
        int i=0,j=height.size()-1;
        int capity = 0;
        int max = min(height[i],height[j])*(j-i);
        while (i<j) {
            if (height[i] < height[j]) {
                capity = height[i]*(j-i);
                i++;
            } else {
                capity = height[j]*(j-i);
                j--;
            }
            // cout << capity << endl;
            max = capity > max ? capity : max;
        }
        return max;
    }
};
int main() {
    Solution solve;
    vector<int> height = {1,1};
    cout << solve.maxArea(height);
}

File: Problem11_ContainerWithMostWater.py
from typing import List

class Solution:
    def maxArea(self, height: List[int]) -> int:
        i=0
        j=len(height)-1
        maxA = 0
        while i<j:
            if height[i] < height[j]:
                area = height[i]*(j-i)
                i+=1
            else:
                area = height[j]*(j-i)
                j-=1
            maxA = maxA if maxA>area else area
        return maxA

if __name__ == "__main__":
    solve = Solution()
    height = [1,1]
    print(solve.maxArea(height))
        

# //i=0, j=n-1
# // nếu h(i) < h(j):(True)
# // h(i)*(j-i) = 1*(8-0) = 8
# // tăng i lên
# // else:
# // h(j)*(j-i) = 7*(8-1) = 49
# // giảm j xuống